pkg/cloud/azure/actuators/machine_scope.go

```diff
@@ -400,3 +418,99 @@ func getEnvironment(m *MachineScope) (*azure.Environment, error) {
+		return *to.StringMapPtr(ocpTags), nil
+	}
+
+	// Tag keys are case-insensitive. A tag with a key, regardless of
+	// the casing, is updated or retrieved. An Azure service might keep
+	// the casing as provided for the tag key. To allow user to choose
+	// the required variant of the key to add. return error when
+	// duplicate tag keys are found in machineSpecTags. Tags in
+	// infraStatusTags is vetted by installer.
+	if err := findDuplicateTagKeys(machineSpecTags); err != nil {
+		return nil, fmt.Errorf("Machine.Spec.ProviderSpec.Tags validation failed: %w", err)
+	}
+
+	tags := make(map[string]*string)
+	// copy user defined tags in Infrastructure.Status.
+	for k, v := range infraStatusTags {
+		tags[k] = to.StringPtr(v)
+	}
+
+	// merge tags present in Infrastructure.Status with
+	// the tags configured in AzureMachineProviderSpec, with
+	// precedence given to those in AzureMachineProviderSpec
+	// for new or updated tags.
+	for k, v := range machineSpecTags {
+		tags[k] = to.StringPtr(v)
+	}
+
+	// copy OCP tags, overwrite any OCP reserved tags found in
+	// the user defined tag list.
+	for k, v := range ocpTags {
+		tags[k] = to.StringPtr(v)
+	}
+
+	// Automation, Content Delivery Network, DNS Azure resources can have a
+	// maximum of 15 tags. OpenShift is reserved with 5 tags for internal use,
+	// allowing 10 tags for user configuration. But this operator does not create
+	// any of the above resources and supports configuring additional tags, hence
+	// increasing the user tags max limit to 45.
+	if len(ocpTags) > 5 || (len(tags)-len(ocpTags)) > 45 {
+		return nil, fmt.Errorf("ocp can define upto 5 tags and user can define upto 45 tags,"+
+			"infrstructure.status.resourceTags and Machine.Spec.ProviderSpec.Tags put together, current tag count: %d", len(tags))
+	}
+
+	return tags, nil
+}
+
+func getInfraResourceTags(platformStatus *configv1.PlatformStatus) (tags map[string]string) {
+	if platformStatus != nil && platformStatus.Azure != nil && platformStatus.Azure.ResourceTags != nil {
+		tags = make(map[string]string, len(platformStatus.Azure.ResourceTags))
+		for _, tag := range platformStatus.Azure.ResourceTags {
+			tags[tag.Key] = tag.Value
+		}
+	}
+	return
+}
+
+func getOCPTagList(clusterID string) (map[string]string, error) {
+	if clusterID == "" {
+		return nil, fmt.Errorf("cluster ID required for generating OCP tag list")
+	}
+	return map[string]string{
+		fmt.Sprintf("kubernetes.io_cluster.%v", clusterID): "owned",
+	}, nil
+}
+
+func findDuplicateTagKeys(tagSet map[string]string) (err error) {
+	if len(tagSet) == 0 {
+		return nil
+	}
+
+	dupKeys := make(map[string]int)
+	for k := range tagSet {
+		dupKeys[strings.ToTitle(k)]++
+	}
+
+	var errMsg []string
+	for key, count := range dupKeys {
+		if count > 1 {
+			errMsg = append(errMsg, fmt.Sprintf("\"%s\" matches %d keys", key, count))
+		}
+	}
+	if len(errMsg) > 0 {
+		err = fmt.Errorf("found duplicate tag keys: %v", strings.Join(errMsg, ", "))
+	}
+
+	return err
+}
```

Comment:
No need to use a named return here, prefer to do this please
```suggestion
func findDuplicateTagKeys(tagSet map[string]string) error {
	if len(tagSet) == 0 {
		return nil
	}

	dupKeys := make(map[string]int)
	for k := range tagSet {
		dupKeys[strings.ToTitle(k)]++
	}

	var errMsg []string
	for key, count := range dupKeys {
		if count > 1 {
			errMsg = append(errMsg, fmt.Sprintf("\"%s\" matches %d keys", key, count))
		}
	}
	if len(errMsg) > 0 {
		return fmt.Errorf("found duplicate tag keys: %v", strings.Join(errMsg, ", "))
	}

	return nil
}
```